lib/kafka/async_producer.rb

```diff
@@ -244,6 +247,11 @@ def do_loop
                 @instrumenter.instrument("drop_messages.async_producer", {
                   message_count: @producer.buffer_size + @queue.size,
                 })
+
+                if !@finally.nil?
```

Comment:
Two thoughts: 
- Without redesigning the whole producer I'm not entirely sure we can do anything about the worker never getting the `shutdown` message.
- If we do make the retries finite then maybe this `:shutdown` operation ends up basically never getting used. It might come at the cost of more messages getting stashed than the "try until you die" strategy.